### PR TITLE
Refactoring HTTP downloader progress reporter to accept multiple observers

### PR DESCRIPTION
**Pull request by ycombinator**
## What does this PR do?

Prior to this PR, the HTTP downloader's `progressReporter` would report download progress in Agent logs. This functionality was implemented entirely within the `progressReporter` code.

This PR refactors the `progressReporter` code to accept multiple observers and introduces a new observer, `loggingProgressObserver`, that logs the observed progress in Agent logs, preserving the behavior prior to this PR. 

## Why is it important?

This PR is purely a refactoring PR; it does not change any functionality.  Functionally, the HTTP downloader's progress will continue to be reported in Agent logs. 

However, in https://github.com/elastic/elastic-agent/pull/3527, we will need the `progressReporter` to also report progress to another observer.  This PR makes it easy to accomplish that in a clean manner.  More context here: https://github.com/elastic/elastic-agent/pull/3527#discussion_r1347016802

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] ~I have made corresponding changes to the documentation~
- [ ] ~I have made corresponding change to the default configuration files~
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] ~I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)~
- [ ] ~I have added an integration test or an E2E test~

## Related issues

- Relates to https://github.com/elastic/elastic-agent/pull/3527#discussion_r1347016802

### CODE REVIEW

**Comment by ycombinator** [comment]
While working on this PR, I found a bug with the progress reporter.  It was reporting progress just once (after a delay) instead of reporting it periodically.  As such, this PR here is now blocked on the bugfix PR: https://github.com/elastic/elastic-agent/pull/3548.

File: internal/pkg/agent/application/upgrade/artifact/download/http/downloader.go
Comment (michalpristas) [inline internal/pkg/agent/application/upgrade/artifact/download/http/downloader.go:209]: wild name, i had to go up here to figure it what it means

File: internal/pkg/agent/application/upgrade/artifact/download/http/progress_reporter.go
Comment (michalpristas) [inline internal/pkg/agent/application/upgrade/artifact/download/http/progress_reporter.go:54]: very unnecessary optimization from my side: no need to spin up goroutine or initiate tickers if progressObservers is []. as we're having at least log now, i consider this just a small optimization from component side of view.

File: internal/pkg/agent/application/upgrade/artifact/download/http/progress_reporter.go
Comment (michalpristas) [inline internal/pkg/agent/application/upgrade/artifact/download/http/progress_reporter.go:59]: we're counting on external elements to cancel this. we could introduce Done chan internally which would be closed on ReportComplete or ReportFailed

**Review by michalpristas — APPROVED** [review]
just a very tiny comments, it looks good otherwise

File: internal/pkg/agent/application/upgrade/artifact/download/http/progress_reporter.go
Comment (michalpristas) [inline internal/pkg/agent/application/upgrade/artifact/download/http/progress_reporter.go:47]: make a comment that called is responsible for cancelling context

File: internal/pkg/agent/application/upgrade/artifact/download/http/downloader.go
Comment (pchila) [inline internal/pkg/agent/application/upgrade/artifact/download/http/downloader.go:49]: Nit: I always prefer the logger as interfaces for mocking and loose coupling, instead of relying on a more concrete type

File: internal/pkg/agent/application/upgrade/artifact/download/http/downloader_test.go
Comment (pchila) [inline internal/pkg/agent/application/upgrade/artifact/download/http/downloader_test.go:240]: If there are other logs mixed in with the progress logs, this assert will fail the test. Is this intentional?

File: internal/pkg/agent/application/upgrade/artifact/download/http/progress_reporter.go
Comment (ycombinator) [reply internal/pkg/agent/application/upgrade/artifact/download/http/progress_reporter.go:54]: Good point, will add, thanks!

File: internal/pkg/agent/application/upgrade/artifact/download/http/progress_reporter.go
Comment (ycombinator) [reply internal/pkg/agent/application/upgrade/artifact/download/http/progress_reporter.go:49]: Done in 89ca86b27d49e1cfa981f9156886689ffc9d1218.

File: internal/pkg/agent/application/upgrade/artifact/download/http/downloader.go
Comment (ycombinator) [reply internal/pkg/agent/application/upgrade/artifact/download/http/downloader.go]: Fixed in e9fc0970bac5927bd135c4f0a78b36b74d1eae5a.

File: internal/pkg/agent/application/upgrade/artifact/download/http/progress_reporter.go
Comment (ycombinator) [reply internal/pkg/agent/application/upgrade/artifact/download/http/progress_reporter.go:61]: Added in da06f8c9b.

File: internal/pkg/agent/application/upgrade/artifact/download/http/progress_reporter.go
Comment (pchila) [inline internal/pkg/agent/application/upgrade/artifact/download/http/progress_reporter.go:46]: Why the downloadProgressReporter needs to implement a `Writer` interface that only increments an integer ?

File: internal/pkg/agent/application/upgrade/artifact/download/http/progress_reporter.go
Comment (ycombinator) [reply internal/pkg/agent/application/upgrade/artifact/download/http/progress_reporter.go]: Implemented your suggestion in 67f80f63f.  Let me know what you think.

Comment (pchila) [reply internal/pkg/agent/application/upgrade/artifact/download/http/progress_reporter.go]: what about creating a subcontext that can be closed when the request is complete or cancelled (ofc we need to hold a ref to its cancellation function) ?

**Review by pchila — COMMENTED** [review]
A few nits, nothing major.
Left a couple of questions...

File: internal/pkg/agent/application/upgrade/artifact/download/http/downloader.go
Comment (ycombinator) [reply internal/pkg/agent/application/upgrade/artifact/download/http/downloader.go:49]: In general, I agree but I'm also a fan of YAGNI.  We can easily change this back to an interface if/when we need it, i.e. have more than one implementation.

File: internal/pkg/agent/application/upgrade/artifact/download/http/downloader_test.go
Comment (ycombinator) [reply internal/pkg/agent/application/upgrade/artifact/download/http/downloader_test.go:240]: Good point.  As things stand, the only other log that could get mixed up is this one: https://github.com/elastic/elastic-agent/blob/359be4f18ef1104f60e9199aca268d5dcbfd77f1/internal/pkg/agent/application/upgrade/artifact/download/http/downloader.go#L102-L104

This log _should_ never get triggered but I suppose in some strange circumstances the `os.Remove` could fail. 

Any suggestions on how to make this more robust to only consider progress logs? Maybe the logs could first be filtered to only keep ones that match `expectedProgressRegexp` or `expectedCompletedRegexp` before running any assertions?

File: internal/pkg/agent/application/upgrade/artifact/download/http/progress_reporter.go
Comment (ycombinator) [reply internal/pkg/agent/application/upgrade/artifact/download/http/progress_reporter.go:46]: This being a refactoring PR, I didn't want to change the implementation too much.  So I didn't touch this.  However, the answer lies in how the progress reporter is informed of a download's progress, over here: https://github.com/elastic/elastic-agent/blob/359be4f18ef1104f60e9199aca268d5dcbfd77f1/internal/pkg/agent/application/upgrade/artifact/download/http/downloader.go#L214

File: internal/pkg/agent/application/upgrade/artifact/download/http/progress_reporter.go
Comment (ycombinator) [reply internal/pkg/agent/application/upgrade/artifact/download/http/progress_reporter.go]: > what about creating a subcontext that can be closed when the request is complete or cancelled (ofc we need to hold a ref to its cancellation function) ?

I like this pattern better as it doesn't require the consumer to remember to necessarily call either `ReportComplete` or `ReportFailed`.  Basically, it gives the consumer two options on how to cancel: either cancel the `ctx` passed to `Report` or call `ReportComplete`/`ReportFailed`. 

As far as implementation goes, we could do a subcontext or we use an internal `done` channel as implemented in https://github.com/elastic/elastic-agent/commit/67f80f63fe8c8359e381bc3233234c99ce4f1c41. Is one implementation necessarily better than the other?

Comment (pchila) [reply internal/pkg/agent/application/upgrade/artifact/download/http/progress_reporter.go]: The context cancellation is more "abandon processing" and the done channel is the simple solution to unblock a go routine.

The context has the advantage of propagation through the subcontexts if the parent context is cancelled/closed/expires, the done channel has to be managed/closed explicitly.

Since you have already implemented the `done` channel you can keep that, I am just quite surprised to see that we only send an empty struct instead of closing the channel and removing the reference

File: internal/pkg/agent/application/upgrade/artifact/download/http/downloader_test.go
Comment (pchila) [reply internal/pkg/agent/application/upgrade/artifact/download/http/downloader_test.go:240]: Filtering and then asserting length is surely one option. The exact semantics depend on what we are trying to assert here: from the code you are asserting that at least the first `minExpectedProgressLogs` log entries must be the ones you expect: this will break as soon as an extra log entry is in the mix (it may even be added in the future).

Did you mean to assert that among the collected log entries you have at least `minExpectedProgressLogs`  expected logs ? In this case filter/count and then assert

Comment (ycombinator) [reply internal/pkg/agent/application/upgrade/artifact/download/http/downloader_test.go:240]: Let me put up a commit showing what I meant by filtering and see what you think.

**Review by blakerouse — APPROVED** [review]
This looks good to me. I tried to find something to complain about, I was unsuccessful. ;-)

File: internal/pkg/agent/application/upgrade/artifact/download/http/progress_reporter.go
Comment (ycombinator) [reply internal/pkg/agent/application/upgrade/artifact/download/http/progress_reporter.go:69]: @pchila @michalpristas I updated the implementation such that there are now two ways to return from the goroutine inside `Report`:
* `Report` is passed a context.  When the context is done, the goroutine will return.
* When the consumer calls `ReportComplete` or `ReportFailed`, we close an internal `done` channel, which will also cause the goroutine in `Report` to return.

File: internal/pkg/agent/application/upgrade/artifact/download/http/downloader_test.go
Comment (ycombinator) [reply internal/pkg/agent/application/upgrade/artifact/download/http/downloader_test.go:252]: @pchila I implemented the filtering in a464077ad.  Let me know what you think.

**Review by pchila — APPROVED** [review]

**Pull request merged**
